### PR TITLE
bug: felt to bytes little underconstrained

### LINKED ISSUE #1300: bug: felt_to_bytes_little is underconstrained
### Kakarot version

The individual `[output]``s are not constrained to be modulo `base`.

https://github.com/kkrt-labs/kakarot/blob/a41fe3b3e2f4d77a81028777e37813775b824e55/src/utils/bytes.cairo#L62

### PR DESCRIPTION
**Pull request by obatirou**
Time spent on this PR:

## Pull request type

Please check the type of change your PR introduces:

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves #1300

## What is the new behavior?

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot/1317)

### CODE REVIEW

File: src/utils/bytes.cairo [hunk 1]
```diff
@@ -60,22 +63,117 @@ func felt_to_bytes_little(dst: felt*, value: felt) -> felt {
...
+    let is_inf_to_max = is_le_felt(bytes_len, max_bytes_used);
+    tempvar range_check_ptr = range_check_ptr;
+    with_attr error_message("bytes_len superior to max_bytes_used") {
+        assert is_inf_to_max = 1;
+    }
+
     return bytes_len;
 }
 
+func bytes_used_128{range_check_ptr}(value: felt) -> felt {
```
Comment (obatirou) [inline src/utils/bytes.cairo:108]: Here due to circular dependencies, should be moved if solution is retained

File: tests/src/utils/test_bytes.py [hunk 2]
```diff
@@ -15,12 +19,44 @@ def test_should_return_ascii(self, cairo_run, n):
...
+                    "memory[ids.output] = (int(ids.value) % PRIME)\n",
+                ),
+                cairo_error(message="output superior to base"),
+            ):
+                cairo_run("test__felt_to_bytes_little", n=n)
+
+        @given(
+            n=integers(
+                min_value=452312848583266388373324160190187140051835877600158453279131187530910663137,
+                max_value=452312848583266388373324160190187140051835877600158453279131187530910663137,
```
Comment (obatirou) [inline tests/src/utils/test_bytes.py:45]: Special case found by hypothesis where the decomposition is not unique

Comment (obatirou) [reply tests/src/utils/test_bytes.py:45]: Assumption taken from split int hint was not followed

File: src/utils/bytes.cairo (hunk 1 above)
Comment (obatirou) [reply src/utils/bytes.cairo]: New solution retained, not relevant anymore

File: src/utils/bytes.cairo [hunk 3]
```diff
@@ -1,5 +1,11 @@
 from starkware.cairo.common.alloc import alloc
-from starkware.cairo.common.math import unsigned_div_rem, split_int, split_felt
+from starkware.cairo.common.math import (
+    unsigned_div_rem,
+    split_int,
+    split_felt,
```
Comment (enitrat) [inline src/utils/bytes.cairo:5]:
```suggestion
```

File: src/utils/bytes.cairo [hunk 4]
```diff
@@ -59,23 +72,80 @@ func felt_to_bytes_little(dst: felt*, value: felt) -> felt {
         memory[ids.output] = res = (int(ids.value) % PRIME) % ids.base
         assert res < ids.bound, f'split_int(): Limb {res} is out of range.'
     %}
-    let byte = [output];
-    let value = (value - byte) / base;
+    tempvar byte = [output];
```
Comment (enitrat) [inline src/utils/bytes.cairo:75]:
```suggestion
    let byte = [output];
```

this one can be a simple `let` reference

File: tests/src/utils/test_bytes.py [hunk 5]
```diff
@@ -15,14 +19,63 @@ def test_should_return_ascii(self, cairo_run, n):
...
+        # The value 0 is also excluded as it is treated as a special case in the function
+        @given(
+            n=integers(min_value=1, max_value=2**248 - 1).filter(
+                lambda x: int.to_bytes(
+                    x, length=(x.bit_length() + 7) // 8, byteorder="little"
+                )[0]
+                != 0
+            )
+        )
+        def test_should_raise_when_bytes_len_is_not_minimal(
```
Comment (enitrat) [inline tests/src/utils/test_bytes.py:64]: great 👏

File: src/utils/bytes.cairo (hunk 3 above)
Comment (obatirou) [reply src/utils/bytes.cairo:5]: Those are used in other functions from the files

**Review by enitrat — DISMISSED** [review]

**Review by enitrat — APPROVED** [review]

**Pull request merged**